### PR TITLE
[9.x] Allow HTTP client requests with retries to optionally throw RequestExceptions

### PR DESCRIPTION
**Pull request by mattkingshott**
This PR addresses #40055 by adding functionality to the HTTP client so that you can prevent it from throwing a `RequestException` when `retries` are configured.

## The problem

Presently, and as highlighted in the docs, the HTTP client does not throw exceptions when a request fails. As a result, you can make use of methods like `failed` on the response e.g.

```php
$response = Http::get('http://example.com');

if ($response->failed())
{
    // Handle the error
}
```

However, if you configure `retries` on the HTTP client, it will throw a `RequestException` if all attempts fail. 

```php
$response = Http::retries(3, 1000)->get('http://example.com');
```

This means:

1. You cannot configure `throw()` on the client.
2. You cannot chain methods e.g. `throw()->json()`.
3. You cannot use `throwIf`.
4. You cannot use the response's error handling methods e.g. `$response->failed()`.

In addition, you have to wrap the request within a try/catch block if you want to handle the error. 

## The solution

The PR adds a fourth parameter `throw` to the `retries` method, which allows you to control whether the HTTP client should throw an exception when the retry limit is reached. 

By default, this parameter is set to `true` to preserve backward compatibility.

```php
Http::retries(3, 1000); // throws a RequestException if all retries fail
Http::retries(3, 1000, null, true); // throws a RequestException if all retries fail
Http::retries(3, 1000, null, false); // does not throw a RequestException if all retries fail
```

### CODE REVIEW

File: src/Illuminate/Http/Client/PendingRequest.php [hunk 1]
```diff
@@ -451,12 +458,14 @@ public function timeout(int $seconds)
      * @param  int  $times
      * @param  int  $sleep
      * @param  callable|null  $when
+     * @param  bool  $throw
      * @return $this
      */
-    public function retry(int $times, int $sleep = 0, ?callable $when = null)
+    public function retry(int $times, int $sleep = 0, ?callable $when = null, bool $throw = true)
```
Comment (driesvints) [inline src/Illuminate/Http/Client/PendingRequest.php:464]: We can't change the method signature on a stable release unfortunately.

Comment (mattkingshott) [reply src/Illuminate/Http/Client/PendingRequest.php:464]: Should I submit this to master then?

Comment (driesvints) [reply src/Illuminate/Http/Client/PendingRequest.php:464]: Yeah, if we can't solve it differently then that would be best.

Comment (mattkingshott) [reply src/Illuminate/Http/Client/PendingRequest.php:464]: Okay. I’ll submit it for master.

**Pull request closed without merging**